onnxruntime/core/providers/coreml/builders/impl/split_op_builder.cc

```diff
@@ -43,55 +46,104 @@ Status SplitOpBuilder::AddToModelBuilderImpl(ModelBuilder& model_builder,
-    Initializer unpacked_tensor(split_tensor);
-    auto split_span = unpacked_tensor.DataAsSpan<uint64_t>();
-    auto split_sizes = split_span.size();
-    num_outputs = narrow<uint64_t>(split_sizes);
-    for (size_t i = 0; i < split_sizes; i++) {
-      coreml_splitnd->add_splitsizes(split_span[i]);
-    }
-  } else if (node.SinceVersion() < 18) {
-    num_outputs = narrow<uint64_t>(node.OutputDefs().size());
-    coreml_splitnd->set_numsplits(num_outputs);
-  } else {
-    // note: for opset 18+ 'num_outputs' is a required attribute
-    num_outputs = narrow<uint64_t>(helper.GetInt64("num_outputs").value());
-    // note: checked in IsOpSupportedImpl that ensures the dim value at splitting axis exists
-    auto split_dim_size = data_shape[HandleNegativeAxis(axis, data_shape.size())];
-    uint64_t chunk_size = narrow<uint64_t>((split_dim_size + num_outputs - 1) / num_outputs);
-    uint64_t remainder = split_dim_size % chunk_size;
-    if (remainder) {
-      // uneven
-      auto split_sizes = InlinedVector<uint64_t>(num_outputs, chunk_size);
-      split_sizes.back() = remainder;
-      for (size_t i = 0; i < split_sizes.size(); i++) {
-        coreml_splitnd->add_splitsizes(split_sizes[i]);
+  int64_t axis = helper.Get("axis", 0);
+
+
+#if defined(COREML_ENABLE_MLPROGRAM)
+  if (model_builder.CreateMLProgram()) {
+      int64_t num_outputs;
+      using namespace CoreML::Specification::MILSpec;
+      std::unique_ptr<Operation> split_op = model_builder.CreateOperation(node, "split");
+      AddOperationInput(*split_op, "axis", model_builder.AddScalarConstant(split_op->type(), "axis", axis));
+
+      if (input_defs.size() > 1) {
+        // if "split" is explicitly provided as an input
+        const auto& split_tensor = *model_builder.GetInitializerTensors().at(input_defs[1]->Name());
+        Initializer unpacked_tensor(split_tensor);
+        auto split_span = unpacked_tensor.DataAsSpan<int64_t>();
+        num_outputs = narrow<int64_t>(split_span.size());
+        AddOperationInput(*split_op, "split_sizes", model_builder.AddConstant(split_op->type(), "split_sizes", split_span));
+      } else if (node.SinceVersion() < 18) {
+        num_outputs = narrow<uint64_t>(node.OutputDefs().size());
+        AddOperationInput(*split_op, "num_splits",
+            model_builder.AddScalarConstant(split_op->type(), "num_splits", num_outputs));
+      } else {
+        // note: for opset 18+ 'num_outputs' is a required attribute
+        num_outputs = narrow<int64_t>(helper.GetInt64("num_outputs").value());
+        // note: checked in IsOpSupportedImpl that ensures the dim value at splitting axis exists
+        auto split_dim_size = data_shape[HandleNegativeAxis(axis, data_shape.size())];
+        uint64_t chunk_size = narrow<int64_t>((split_dim_size + num_outputs - 1) / num_outputs);
+        uint64_t remainder = split_dim_size % chunk_size;
+
+        if (remainder) {
```

Comment:
The code to calculate split_sizes if opset 18+ and num_outputs results in an uneven split is the same for NN and ML Program isn't it? If so we could move it up so it's not duplicated? 

In the shared code it may be simplest to narrow to int32_t when calculating the sizes for an uneven split so they're easily used by NN (uint64) and MLProgram (AddConstant takes int64_t). I don't believe a valid split will ever have a value larger than the max value of int32_t.

Reply:
I created inline function to reduce duplicate code. But there might be better way to restructure the code as well. I am wondering if having separate function for MLProgram and Neural Network might be better in future.

Reply:
inline function is fine.

Sorry - I should have explained a bit more. the idea behind using int32_t for the calculations was that we'd just need to do a `narrow` once and return a 32-bit data type that could be converted to int64_t for the ML Program input or uint64_t for the NN input without requiring any more `narrow` statements. not a big deal - but a narrow does a check and potentially throws which adds a little bit of binary size.

Reply:
Regarding a separate function for NN vs MLProgram: Binary size is a big consideration in mobile, so if there's an overlap in the code required for each implementation we need to be careful about duplicating code unnecessarily. That's probably the main consideration for deciding whether a separate function is a better option.